File: quiz.lua
-- Dependencies
local function ensureHtmlDeps()
  quarto.doc.addHtmlDependency({
    name = "quiz",
    version = "1.0.0",
    scripts = {"quiz.js"},
    stylesheets = {"quiz.css"}
  })
end

-- Escape function for JSON attributes
local function escapeJSON(s)
  if s == nil then return "" end
  s = s:gsub('\\', '\\\\')
  s = s:gsub('"', '\\"')
  s = s:gsub('\n', '\\n')
  s = s:gsub('\r', '\\r')
  s = s:gsub('\t', '\\t')
  return s
end

-- Function to get environment variable or fallback to the provided value
local function getEnvOrValue(varName, defaultValue)
  local value = os.getenv(varName)
  if value and value ~= "" then
    return value
  else
    return defaultValue
  end
end

-- Main filter
function Div(el)
  if el.classes:includes("quiz") then
    ensureHtmlDeps()
    
    local title = el.attributes["title"] or "Untitled Quiz"
    local description = el.attributes["description"] or ""
    
    -- Check for API key in attributes or environment variable
    local api_key = el.attributes["api-key"] or ""
    local env_var_name = el.attributes["api-key-env"] or "QUIZ_API_KEY"
    
    -- If api-key attribute is not provided or empty, try to get from environment variable
    if api_key == "" then
      api_key = getEnvOrValue(env_var_name, "")
    end
    
    local quiz_div = pandoc.Div({
      pandoc.RawBlock('html', [[
        <script src="https://cdnjs.cloudflare.com/ajax/libs/qrcode-generator/1.4.4/qrcode.min.js"></script>
      ]]),
      pandoc.RawBlock('html', string.format([[
        <div class="quiz-container" 
             data-title="%s" 
             data-description="%s"
             data-api-key="%s"
             data-api-key-env="%s">
          <div class="quiz-header">
            <h3>%s</h3>
            <p class="quiz-description">%s</p>
            <button class="create-quiz">Create Quiz Session</button>
          </div>
          <div class="quiz-content" style="display: none;">
            <div class="qr-section">
              <div class="qr-container"></div>
              <div class="qr-info">
                <p class="qr-title">Join the quiz!</p>
                <p class="qr-subtitle">Scan the QR code or click the link below</p>
                <div class="qr-link-container"></div>
              </div>
            </div>
          </div>
        </div>
      ]], title, description, escapeJSON(api_key), escapeJSON(env_var_name), title, description))
    })
    
    return quiz_div
    
  elseif el.classes:includes("question") then
    ensureHtmlDeps()
    
    -- Step 1: Separate the content into question and options parts
    local questionBlocks = {}
    local bulletListIndex = nil
    
    for i, block in ipairs(el.content) do
      if block.t == "BulletList" then
        bulletListIndex = i
        break
      else
        table.insert(questionBlocks, block)
      end
    end
    
    -- Step 2: Process the question part
    local questionPandoc = pandoc.Pandoc(questionBlocks)
    local questionHtml = pandoc.write(questionPandoc, "html")
    
    -- Create a plain text version for the data attribute
    local questionPlainText = ""
    if #questionBlocks > 0 then
      questionPlainText = pandoc.utils.stringify(pandoc.Pandoc(questionBlocks))
    end
    
    -- Step 3: Process the options (BulletList)
    local options = {}
    local optionsHtml = ""
    
    if bulletListIndex then
      local bulletList = el.content[bulletListIndex]
      
      for i, item in ipairs(bulletList.content) do
        -- Get the first paragraph of the list item
        local firstPara = item[1]
        local isCorrect = false
        local optionContent = {}
        
        -- Check if there's a checkbox
        if firstPara and firstPara.content and #firstPara.content > 0 then
          local firstInline = firstPara.content[1]
          
          -- Check for checkbox in the text
          if firstInline.text then
            local text = firstInline.text
            isCorrect = text:match("%[x%]") or text:match("☒")
            
            -- Create a copy of the paragraph content for processing
            local newContent = pandoc.List()
            local skipFirst = false
            
            -- Check and remove the checkbox marker
            if text:match("^%s*-%s*%[.%]%s*") or text:match("^%s*☐%s*") or text:match("^%s*☒%s*") then
              -- Remove checkbox from text
              local newText = text:gsub("^%s*-%s*%[.%]%s*", ""):gsub("^%s*☐%s*", ""):gsub("^%s*☒%s*", "")
              
              if newText ~= "" then
                -- Replace the first inline with modified text
                local newFirstInline = firstInline:clone()
                newFirstInline.text = newText
                newContent:insert(newFirstInline)
              else
                -- Skip empty text node
                skipFirst = true
              end
              
              -- Add rest of content
              for j = 2, #firstPara.content do
                newContent:insert(firstPara.content[j])
              end
            else
              -- No checkbox found, keep content as is
              newContent = firstPara.content
            end
            
            -- Create new paragraph with processed content
            local newPara = pandoc.Para(newContent)
            
            -- Create a new list item with processed content
            local newItem = pandoc.List()
            newItem:insert(newPara)
            
            -- Copy any additional blocks from the original item
            for j = 2, #item do
              newItem:insert(item[j])
            end
            
            -- Get the original markdown by converting to markdown
            local optMarkdown = pandoc.write(pandoc.Pandoc(newItem), "markdown")
            -- Clean up markdown (remove some extra newlines, etc.)
            optMarkdown = optMarkdown:gsub("^\n+", ""):gsub("\n+$", "")
            
            -- Create a plain text version for display attributes
            local optPlainText = pandoc.utils.stringify(pandoc.Pandoc({pandoc.Plain(newContent)}))
            
            -- Create HTML version of the option content for display
            local optHtml = pandoc.write(pandoc.Pandoc(newItem), "html")
            -- Clean up HTML to remove unnecessary paragraph tags
            optHtml = optHtml:gsub("^%s*<p>", ""):gsub("</p>%s*$", "")
            
            -- Add to options list, including both plain text and markdown
            table.insert(options, {
              text = escapeJSON(optPlainText),
              markdown = escapeJSON(optMarkdown),
              html = optHtml,
              is_correct = isCorrect
            })
          end
        end
      end
    end
    
    -- Build options JSON, including markdown content
    local optionsJson = "["
    for i, opt in ipairs(options) do
      if i > 1 then optionsJson = optionsJson .. "," end
      optionsJson = optionsJson .. string.format(
        '{"text":"%s","markdown":"%s","is_correct":%s}',
        opt.text,
        opt.markdown,
        opt.is_correct and "true" or "false"
      )
    end
    optionsJson = optionsJson .. "]"
    
    -- Build options HTML
    local optionsHtmlStr = ""
    for i, opt in ipairs(options) do
      optionsHtmlStr = optionsHtmlStr .. string.format([[
        <div class="option" data-index="%d">
          <div class="option-text">%s</div>
          <span class="option-stats"></span>
        </div>
      ]], i-1, opt.html)
    end
    
    -- Create HTML
    local html = string.format([[
      <div class="question-container" 
           data-question="%s"
           data-question-markdown="%s"
           data-options='%s'>
        <div class="question-text">%s</div>
        <div class="options-display">%s</div>
        <div class="options-container"></div>
        <div class="question-status">
          <span class="answer-count" style="display: none;">0 answers</span>
        </div>
        <div class="question-buttons">
          <button class="submit-question">Present Question</button>
          <button class="close-question" style="display: none;">Close Question</button>
        </div>
      </div>
    ]], escapeJSON(questionPlainText), escapeJSON(questionMarkdown), optionsJson, questionHtml, optionsHtmlStr)
    
    return pandoc.RawBlock('html', html)
  end
end

File: quiz.js
// Configuration
const CONFIG = {
    BACKEND_URL: 'https://quiz-back-t8sa.onrender.com',
    STUDENT_URL: 'https://quiz-front-61g6.onrender.com'
};

document.addEventListener('DOMContentLoaded', function () {
    // Handle quiz containers
    const quizContainers = document.querySelectorAll('.quiz-container');

    quizContainers.forEach(function (container) {
        const createButton = container.querySelector('.create-quiz');
        const quizContent = container.querySelector('.quiz-content');
        const qrContainer = container.querySelector('.qr-container');
        const qrLinkContainer = container.querySelector('.qr-link-container');
        const title = container.dataset.title;
        const description = container.dataset.description;
        
        // Get API key from data attribute
        const apiKey = container.dataset.apiKey;
        const apiKeyEnv = container.dataset.apiKeyEnv;

        // Check if API key is provided
        if (!apiKey) {
            console.error('No API key provided. Please either:');
            console.error('1. Add an api-key attribute to your quiz element, or');
            console.error(`2. Set the ${apiKeyEnv} environment variable.`);
            
            createButton.disabled = true;
            
            // Create API key input form
            const keyForm = document.createElement('div');
            keyForm.className = 'api-key-form';
            keyForm.innerHTML = `
                <h3>API Key Required</h3>
                <p>Please enter your API key to enable the quiz functionality:</p>
                <div class="input-group">
                    <input type="password" class="api-key-input" placeholder="Enter your API key">
                    <button class="save-api-key">Save</button>
                </div>
            `;
            
            // Insert the form before the create button
            container.querySelector('.quiz-header').insertBefore(keyForm, createButton);
            
            // Handle save button click
            const saveButton = keyForm.querySelector('.save-api-key');
            const keyInput = keyForm.querySelector('.api-key-input');
            
            saveButton.addEventListener('click', function() {
                const enteredKey = keyInput.value.trim();
                if (enteredKey) {
                    // Set the key as a data attribute and enable the create button
                    container.dataset.apiKey = enteredKey;
                    createButton.disabled = false;
                    keyForm.style.display = 'none';
                }
            });
            
            return;
        }

        // Function to make authenticated API calls
        async function apiCall(endpoint, options = {}) {
            const headers = {
                'Content-Type': 'application/json',
                'api-key': apiKey, // Use the key from the container
                ...options.headers
            };

            return fetch(`${CONFIG.BACKEND_URL}${endpoint}`, {
                ...options,
                headers
            });
        }

        createButton.addEventListener('click', async function () {
            createButton.disabled = true;
            createButton.textContent = 'Creating quiz session...';

            try {
                // Create new page on the server
                const response = await apiCall('/api/pages/', {
                    method: 'POST',
                    body: JSON.stringify({
                        title: title,
                        description: description
                    })
                });

                if (!response.ok) {
                    throw new Error(`Failed to create quiz page`);
                }

                const data = await response.json();
                const pageId = data.page_id;
                const joinUrl = `${CONFIG.STUDENT_URL}/?page=${pageId}`;

                // Generate QR code
                const qr = qrcode(0, 'L');
                qr.addData(joinUrl);
                qr.make();

                // Create and add QR code image
                const qrImage = document.createElement('div');
                qrImage.innerHTML = qr.createImgTag(5);
                qrContainer.appendChild(qrImage);

                // Display join URL as clickable link
                const urlLink = document.createElement('a');
                urlLink.href = joinUrl;
                urlLink.innerHTML = `
                    <svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
                        <path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path>
                        <polyline points="15 3 21 3 21 9"></polyline>
                        <line x1="10" y1="14" x2="21" y2="3"></line>
                    </svg>
                    Open quiz page
                `;
                urlLink.target = '_blank';
                qrLinkContainer.appendChild(urlLink);

                // Show quiz content
                quizContent.style.display = 'block';
                createButton.textContent = 'Quiz session created!';

                // Store the page ID in a global variable for questions to use
                window.currentQuizPageId = pageId;
                // Also store the API key for questions to use
                window.currentQuizApiKey = apiKey;

            } catch (error) {
                console.error('Error setting up quiz:', error);
                createButton.disabled = false;
                createButton.textContent = 'Create Quiz Session';
            }
        });
    });

    // Handle question containers
    const questionContainers = document.querySelectorAll('.question-container');

    questionContainers.forEach(function (container) {
        const submitButton = container.querySelector('.submit-question');
        const closeButton = container.querySelector('.close-question');
        const answerCount = container.querySelector('.answer-count');
        const questionText = container.dataset.question;
        const questionMarkdown = container.dataset.questionMarkdown || questionText; // Always prefer markdown version
        const options = JSON.parse(container.dataset.options);
        let pollInterval;

        // Function to make authenticated API calls
        async function apiCall(endpoint, options = {}) {
            // Use the API key from the global variable (set when quiz is created)
            if (!window.currentQuizApiKey) {
                console.error('No API key available. Please ensure the quiz is created first.');
                return;
            }

            const headers = {
                'Content-Type': 'application/json',
                'api-key': window.currentQuizApiKey,
                ...options.headers
            };

            return fetch(`${CONFIG.BACKEND_URL}${endpoint}`, {
                ...options,
                headers
            });
        }

        // Function to update answer count
        async function updateAnswerCount() {
            try {
                const response = await apiCall(`/api/pages/${window.currentQuizPageId}`);
                if (!response.ok) {
                    throw new Error('Failed to fetch page status');
                }

                const data = await response.json();
                const count = data.answers ? data.answers.length : 0;
                answerCount.textContent = `${count} ${count === 1 ? 'answer' : 'answers'}`;
            } catch (error) {
                console.error('Error fetching answer count:', error);
            }
        }

        // When submit button is clicked
        submitButton.addEventListener('click', async () => {
            if (!window.currentQuizPageId) {
                console.error('No active quiz page');
                return;
            }

            try {
                // Create a modified options array that uses markdown content
                const optionsWithMarkdown = options.map(option => ({
                    text: option.markdown || option.text, // Use markdown if available
                    is_correct: option.is_correct
                }));
                
                // Post the question to the server
                const response = await apiCall(`/api/pages/${window.currentQuizPageId}/questions`, {
                    method: 'POST',
                    body: JSON.stringify({
                        text: questionMarkdown, // Use markdown version
                        options: optionsWithMarkdown
                    })
                });

                if (!response.ok) {
                    throw new Error('Failed to post question');
                }

                // Show close button and answer count, disable submit button
                submitButton.disabled = true;
                closeButton.style.display = 'block';
                answerCount.style.display = 'inline-block';

                // Start polling for answer count
                updateAnswerCount();
                pollInterval = setInterval(updateAnswerCount, 2000);

            } catch (error) {
                console.error('Error submitting question:', error);
            }
        });

        // When close button is clicked
        closeButton.addEventListener('click', async () => {
            try {
                // Stop polling
                clearInterval(pollInterval);

                // Close the question via API
                const response = await apiCall(`/api/pages/${window.currentQuizPageId}/close-question`, {
                    method: 'POST'
                });

                if (!response.ok) {
                    throw new Error('Failed to close question');
                }

                const stats = await response.json();

                // Update each option with stats
                for (const [index, opt] of options.entries()) {
                    const optionStats = stats.option_stats[index];
                    const optionEl = container.querySelector(`.option[data-index="${index}"]`);
                    const statsEl = optionEl.querySelector('.option-stats');

                    // Update stats text
                    statsEl.textContent = `(${optionStats.count} / ${optionStats.percentage.toFixed(0)}%)`;

                    // Update styling
                    if (opt.is_correct) {
                        optionEl.classList.add('correct');
                    } else {
                        optionEl.classList.add('incorrect');
                    }
                }

                // Hide answer count and disable close button
                answerCount.style.display = 'none';
                closeButton.disabled = true;
                closeButton.textContent = 'Question Closed';

            } catch (error) {
                console.error('Error closing question:', error);
            }
        });
    });
});